scipy/linalg/_matfuncs_expm.c

```diff
@@ -0,0 +1,2375 @@
+#include "_matfuncs_expm.h"
+
+// For some reason these defs are not picked up from the header.
+#if defined(_MSC_VER)
+    // MSVC nonsense
+    #include <complex.h>
+    #define EXPM_Z _Dcomplex
+    #define EXPM_C _Fcomplex
+    #define CPLX_Z(real, imag) (_Cbuild(real, imag))
+    #define CPLX_C(real, imag) (_FCbuild(real, imag))
+#else
+    // C99-compliant compilers
+    #include <complex.h>
+    #define EXPM_Z double complex
+    #define EXPM_C float complex
+    #define CPLX_Z(real, imag) (real + imag*I)
+    #define CPLX_C(real, imag) (real + imag*I)
+#endif
+
+
+static float snorm1(float*, float*, const Py_ssize_t);
+static double dnorm1(double*, double*, const Py_ssize_t);
+static float cnorm1(EXPM_C*, float*, const Py_ssize_t);
```

Comment:
These s- d- c- z- variants really look like c++ template material (or back to temporary heh). Also complex math might be a bit saner

Reply:
Quite possible. I tried a little bit but then I gave up. I am hoping that anyone can grab these and put them into their project outside Python including C. And there are enough annoyances with if else conditions depending on types that makes it ugly again (if `float complex` make the scalar variable `float` but `double` if it is `double complex` but now make it conjugate instead of transpose if it is `double complex` or `float complex` etc.) Then you are in the ugly parts of C++ back again. I think DRY principle is not suitable here.  

I tried a few variants but spelling it out everything and some regex magic turned out to be the easiest option. It's the MSVC that is creating this hopeless situation. Otherwise it is pretty much readable linalg throughout.

Reply:
I have never used that, but would [tgmath.h](https://en.cppreference.com/w/c/numeric/tgmath) help reduce the code duplication? It is defined in C99 and refined in C11. Recent MSVC appears to [support it](https://learn.microsoft.com/en-us/cpp/c-runtime-library/tgmath?view=msvc-170).

Reply:
It's not the functions but arithmetic operators that are missing unfortunately. So `a = b + c` is not working if they are complex. But it's not hard just annoying.